### PR TITLE
feat(session-recording): store session recording data outside of ClickHouse

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

A high proportion of ClickHouse disk space is dedicated to session recordings. These disks are expensive and ClickHouse isn't needed to query the payload data that is using up the space

see #2142 

## Changes

recordings-no-clickhouse

* adds Mini IO to expose S3like storage
* changes the plugin server to write session recording data to disk, not to ClickHouse
* reads the session recording data from disk when it isn't present in ClickHouse
* checks for access to storage when starting the plugin server

## How does it work?

```mermaid
sequenceDiagram
    actor web
    actor SDK
    actor app.posthog.com
    actor kafka
    actor plugin.server
    actor clickhouse
    actor object_storage
    SDK->>app.posthog.com: session_recording_event
    app.posthog.com->>kafka: chunked_session_recording_events
    kafka->>+plugin.server: 
    plugin.server->>clickhouse: stores chunks without data
    plugin.server->>-object_storage: stores chunk data
    web->>+app.posthog.com: load session data
    app.posthog.com->>clickhouse: load session json
    loop for every loaded chunk
        app.posthog.com->>object_storage: load chunk data for session json
    end
    app.posthog.com->>-web: return session data
```

## How did you test this code?

running it locally, adding some tests

### storage becomes mandatory for self-host

Do we support both? Or make it super easy to use storage

We'll need storage for:

 * CSV export
 * image generation
 * potentially ClickHouse backups

### Related PRs
* https://github.com/PostHog/charts-clickhouse/pull/341/
* https://github.com/PostHog/posthog.com/pull/3263

### CODE REVIEW

File: posthog/storage/object_storage.py [hunk 1]
```diff
@@ -30,3 +33,19 @@ def read(file_name: str):
...
+
+def list_all():
+    # TODO page these instead of loading them all
+    return s3.Bucket("posthog").objects.all()
+
+
+def delete(older_than: datetime.datetime):
+    date_limit = pytz.UTC.localize(older_than)
+    count = 0
+    for object in list_all():
```
Comment (guidoiaquinti) [inline posthog/storage/object_storage.py:46]: Nit: a possible scaling issue of doing list_all -> `s3.Bucket("posthog").objects.all()` is that the list could be technically unbounded and return a lot of data. Not sure if the SDK supports it but if yes we could technically something like `s3.Bucket("posthog").objects.filter(last_modified<date_limit).all()`

File: posthog/celery.py [hunk 2]
```diff
@@ -370,3 +379,17 @@ def check_async_migration_health():
...
+    """
+    The clickhouse table has a TTL of a number of weeks before session recordings are deleted.
+    Once they are deleted the files they referenced can be deleted
+    """
+    from posthog.internal_metrics import gauge
+
+    ttl_weeks = CONSTANCE_CONFIG["RECORDINGS_TTL_WEEKS"]
+    file_deletion_time_delta = datetime.timedelta(weeks=ttl_weeks, days=1)
+    number_of_deletions = object_storage.delete(datetime.datetime.now() - file_deletion_time_delta)
+    gauge("posthog_celery_session_recordings_deletion", number_of_deletions)
```
Comment (guidoiaquinti) [inline posthog/celery.py:395]: Nit: can we add also a log line when this starts/end and eventual failures?

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 3]
```diff
@@ -712,6 +718,30 @@ export class EventsProcessor {
...
+        const dateKey = castTimestampOrNow(timestamp, TimestampFormat.DateOnly)
+        const object_storage_path = `${OBJECT_STORAGE_SESSION_RECORDING_BUCKET}/${dateKey}/${session_id}/${snapshot_data.chunk_id}/${snapshot_data.chunk_index}`
+        const params = { Bucket: 'posthog', Key: object_storage_path, Body: snapshot_data.data }
+
+        const tags = {
+            team_id: team_id.toString(),
+            session_id,
+        }
+        // TODO: error handling
+        // TODO store path on event
```
Comment (guidoiaquinti) [inline plugin-server/src/worker/ingestion/process-event.ts:733]: outdated comment now that I see line 743?

**Comment by pauldambra** [comment]
@guidoiaquinti 

I'm thinking the order of changes is:

- merge this so it has no effect (helm unchanged, docker doesn't start object storage, code defaults to disabled)
- turn it on on cloud (figuring out what documentation updates will be needed)
    - which is when we can start to save on disk space
- make it opt-in for self-hosted
- get friendly customers to test it
- use that to update documentation
- general release (still opt-in)

(what am I missing/getting wrong :))

**Pull request closed without merging**